From c50145918e4c3fc7838da63451933a7d0d5585e6 Mon Sep 17 00:00:00 2001
From: Michel JAOUEN <michel.jaouen@st.com>
Date: Mon, 10 Feb 2025 15:56:57 +0100
Subject: [PATCH 1563/1695] drivers: stm32_rng: add a PSA call in M33TD CID
 config

On some M33TD CID config (i.e stm32mp2xx) rng IP is not accessible
by optee. A psa call is used.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 2bad82c20cd671ed8cadebe7b7e42aac62f49e63 drivers: stm32_rng: add a PSA call in M33TD CID config]
Change-Id: I0eb961678294a4bb2e81d3bc51636b2509b35741
Signed-off-by: Michel JAOUEN <michel.jaouen@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/437597
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 core/drivers/stm32_rng.c | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/core/drivers/stm32_rng.c b/core/drivers/stm32_rng.c
index 85366ae47..4c7727d57 100644
--- a/core/drivers/stm32_rng.c
+++ b/core/drivers/stm32_rng.c
@@ -22,6 +22,7 @@
 #include <kernel/thread.h>
 #include <libfdt.h>
 #include <mm/core_memprot.h>
+#include <psa/crypto.h>
 #include <rng_support.h>
 #include <stdbool.h>
 #include <stm32_util.h>
@@ -472,14 +473,22 @@ void plat_rng_init(void)
 	uint8_t seed[RNG_FIFO_BYTE_DEPTH] = { };
 
 	if (!stm32_rng) {
-		__plat_rng_init();
-		DMSG("PRNG seeded without RNG");
-		return;
-	}
+		if (IS_ENABLED(CFG_STM32_PSA_SERVICE)) {
 
+			if (psa_generate_random(seed, sizeof(seed)) !=
+			    PSA_SUCCESS)
+				panic();
+			goto init;
+		} else {
+			__plat_rng_init();
+			DMSG("PRNG seeded without RNG");
+			return;
+		}
+	}
 	if (stm32_rng_read(seed, sizeof(seed)))
 		panic();
 
+init:
 	if (crypto_rng_init(seed, sizeof(seed)))
 		panic();
 
-- 
2.47.2

